### ISSUE TITLE
Combine multie checks into one result

### ISSUE THREAD

**Issue by masgo** [centreon/centreon-plugins#630]
This is a feature request. 

I would like to be able to combine the result of multiple checks into one result. 

Examples where this might be useful are: 

- checking traffic for multiple (but not all) interfaces or ports of one device
- fusing sensor data, e.g. temperature, from multiple sensors which are within the same room
- checking (special or rare) OIDs via SNMP which belong together (e.g., traffic up/down or printed pages in black&white + colored + total)
- etc.

The Idea would be to have a plugin, let's call it `combine_multi_checks`, which takes multiple `command` arguments to execute and some options. The commands are all valid checks which reply with a typical plugin response consisting of STATE text | perfdata.

The plugin might have these options:

- `--calc-state [best/worst]` which determines if the overall state is the same as the best or worst encountered state. i.e. worst = a single non-OK response will result in non-OK state / best = a singe OK state is enough to yield an overall OK response
- maybe an option to override the individual warning/critical values?

**Comment by Sims24** [comment]
Hi @masgo 

Thanks for the feedback. FYI I'll work soon on this kind of thing. The "best/worst" is an intersting solution. I need to think about it. 

On my side, here is as I have imagined the feature/plugin after user needs study and thinking plugins as flexible as possible : 

- Query database to gather actual output and state of selected/wanted services directly in real-time databases ;
         --> Avoid executing multiple checks that are already executed by the scheduler on a regular basis ;
         --> Avoid to modify actual behaviour/design of centreon-plugins code (related to exit at the end of a plugin execution) ;
         --> Problem is if you matter about check order, but we could find a way to manage this without re-exec.

- Trigger final state based upon number of ok/warning/critical state that we will add as global perfdata 

- Give ability to make some groups within the check, e.g make a group router where you match all host/services that match a criteria (or a list), and an other group switch etc etc ... with subdivised perfdata on number of ok/warning/critical.

- JSON config file as it has been designed for virtualservice mode and to avoid infinite command-line arguments. 

Concerning the perfdata, it's a little bit harder to manage, because we will probably have some duplicates between checks and not sure it's relevant as we want to consolidate state. That also open a wide and deep discussion on how name it etc. etc. 

Is the described bullet points could answer your need ? Could you imagine using such kind of option definitions for thresholds (keep in mind that will probably be part of JSON example) :

--warning-global '%{count_warning} > 30 && %{count_critical} < 10' --critical-global '%{count_critical} < 10'

Thanks again for your feedback, 
Regards

**Comment by masgo** [comment]
> Query database to gather actual output and state of selected/wanted services directly in real-time databases ;
--> Avoid executing multiple checks that are already executed by the scheduler on a regular basis ;
--> Avoid to modify actual behaviour/design of centreon-plugins code (related to exit at the end of a plugin execution) ;
--> Problem is if you matter about check order, but we could find a way to manage this without re-exec.

My original thought was that the checks I combine do not already exist. But it would be okay to have them first and combine their results later. The only problem might be the time of execution. With my way the combined checking routine would execute the sub-checks one after another. The results would then be within a very short time and therefore they would correlate. (e.g., the sensor data would all be from the same point in time - give or take 1 second) - When you combine individual checks - which are scheduled individually - you might end up combining data which was recorded 60 seconds apart (or even longer for longer checking intervals). But your suggestion would still be better than the status quo.

> Trigger final state based upon number of ok/warning/critical state that we will add as global perfdata

good idea!

> Give ability to make some groups within the check, e.g make a group router where you match all host/services that match a criteria (or a list), and an other group switch etc etc ... with subdivised perfdata on number of ok/warning/critical.

I think this is a case somewhere between a great function and total useless overkill. Might be very nice if it is usable, but might also be quite bad. Look, for example, at the current state of Meta-Services. You can do some nice things with them, but if you have a simple scenario like:  10 locations where each has 2 WAN connections and you just want to combine the traffic graphs per location. You need A LOT of clicks, and in the end you can not have a nice graph which shows the combined up AND down rate in one graph (similar to the default graphs where downloads are visualized as negative values).

> JSON config file as it has been designed for virtualservice mode and to avoid infinite command-line arguments.

This is again quite on the edge between good and bad. One of THE big advantages of Centreon is that you can do (nearly) everything from the web UI. You no longer have to manage config files (like in Nagios). You can define your commands within the UI, link them to templates and services etc. - all from the UI. So, the need to write a JSON config might break that comfort.


Overall, this remembers me of C++ ... if you have well written code by a good and experienced programmer, things are really great - if you have code done by a newbie, you can start to count the WTF/minute at the code review.

**Comment by Sims24** [comment]
Hi @masgo,

Thanks for such great feedback. It's always interesting to have the feeling from end user. 

> The only problem might be the time of execution. With my way the combined checking routine would execute the sub-checks one after another. The results would then be within a very short time and therefore they would correlate.

A point for you on this. As soon as @garnier-quentin is back, I'll discuss this point with him and we'll take a decision.

>  Look, for example, at the current state of Meta-Services.

Yes, I agree with this, that's why I recently wrote a plugin to mix metric, create some kind of virtual curves, can be several, based on pattern split (traffic_in / traffic_out e.g ;). See https://github.com/centreon/centreon-plugins/pull/593 

> This is again quite on the edge between good and bad. One of THE big advantages of Centreon is that you can do (nearly) everything from the web UI.

Yes, it would be nice to have everything in the UI, but as for virtual service, we could imagine some kind of minimal module to configure and write this JSON, I'm actually trying to learn few thing on web design to be able to do it on my own but I miss some time ... :)

Thanks,

**Comment by masgo** [comment]
@Sims24 

I gave the "using JSON for the config" part another thought; maybe it is not that bad. I realized that there are already plugins which use/support JSON which can be fed to them as command-line parameter (not a file path, but the JSON itself). This can already be done in the UI when defining a check command. 

Okay, it would make the command line reaaaaly long, but no one would have to manually place a file somewhere on the server. This is (probably) also better when dealing with distributed setups where plugins are executed on multiple machines.
